### ISSUE TITLE
Support apt-mark hold

### ISSUE THREAD

**Issue by candlerb** [ansible/ansible#18889]
_From @scottnonnenberg on September 29, 2014 20:37_

##### ISSUE TYPE

Feature Idea
##### COMPONENT NAME

apt
##### ANSIBLE VERSION

N/A
##### SUMMARY

This issue had reasonable interest, and all the `apt`-related modules are Core Modules, so opening a bug here. Here's the original issue: https://github.com/ansible/ansible/issues/8060

_Copied from original issue: ansible/ansible-modules-core#19_

**Comment by candlerb** [comment]
_From @sidharthsurana on September 29, 2014 20:41_

+1

**Comment by candlerb** [comment]
_From @mpdehaan on October 1, 2014 15:30_

Can you please copy the description in to the new ticket?

Thanks!

**Comment by candlerb** [comment]
_From @sidharthsurana on October 1, 2014 15:39_

## Summary from https://github.com/ansible/ansible/issues/8060

Support apt-mark hold/unhold to mark a package as ineligible for automatic updates. Otherwise, ansible can install a particular version of a package, only to see it blown away with a more recent one later.

https://help.ubuntu.com/community/PinningHowto#Introduction%20to%20Holding%20Packages

Steps To Reproduce:
- action: apt pkg=condor=8.0.7-250355 state=present
  ... later
- action: apt upgrade=dist
  ends up with an unexpected version of the package installed.

Expected Results:

Would like to be able to do something like this:
- action: apt pkg=condor=8.0.7-250355 state=present hold=yes
  hold=yes/no would correspond to apt-mark hold/unhold.

For completeness, you could support the manual/auto flag too. So another syntax would be
- action: apt pkg=condor=8.0.7-250355 state=present mark=hold
  ... or ...
- action: apt pkg=condor=8.0.7-250355 state=present mark=hold,manual
  Arguably, whenever you use ansible to install a specific version of a package then it should be held (and implicitly unheld if you specify "state=latest").

ansible 1.4 doesn't appear to do that, but maybe it's changed in later version.

Actual Results:

Currently no obvious ansible way to mark a package as held.

Workaround: use shell to run "apt-mark hold ". But this will be run on every ansible run, and relies on apt-mark itself being idempotent.

**Comment by candlerb** [comment]
_From @bobobox on November 3, 2014 21:46_

This would be nice to have, especially as an older version of a package from a Linux repo probably can't easily be re-installed if inadvertently upgraded.

**Comment by candlerb** [comment]
Probably the right solution is as follows.
- Whenever you install a package with a specific version and `state=present`, the hold flag is automatically set
- Conversely, if you set `state=latest` then the hold flag is automatically cleared.

This involves no additional module options, but just "does the right thing".

Additional data point. At the moment, if the hold flag is set, and you specifically request a particular (different) version in a playbook, ansible aborts with the following error:

```
TASK: [htcondor | apt pkg=condor=8.2.4-281588 state=present] ******************
failed: [host.int.example.net] => {"failed": true}
stderr: E: There are problems and -y was used without --force-yes

stdout: Reading package lists...
Building dependency tree...
Reading state information...
The following packages were automatically installed and are no longer required:
  libaudio-scrobbler-perl libconfig-inifiles-perl linux-headers-3.2.0-32
  linux-headers-3.2.0-29 linux-headers-3.2.0-29-generic
  linux-headers-3.2.0-32-generic libssl0.9.8
Use 'apt-get autoremove' to remove them.
The following held packages will be changed:
  condor
The following packages will be upgraded:
  condor
1 upgraded, 0 newly installed, 0 to remove and 94 not upgraded.
Need to get 32.1 MB of archives.
After this operation, 4680 kB of additional disk space will be used.

msg: 'apt-get install 'condor=8.2.4-281588' ' failed: E: There are problems and -y was used without --force-yes


FATAL: all hosts have already failed -- aborting
```

So it looks like `--force-yes` needs to be provided by ansible if the package is held.

The playbook workaround I'm using for now is this ugly and non-idempotent approach:

```
- action:  shell apt-mark unhold condor
  ignore_errors: yes

- action: apt pkg=condor=8.2.4-281588 state=present

- action:  shell apt-mark hold condor
  ignore_errors: yes
```

**Comment by candlerb** [comment]
_From @kosli on August 16, 2015 22:5_

+1

**Comment by candlerb** [comment]
_From @jakommo on September 7, 2015 10:8_

+1

**Comment by candlerb** [comment]
_From @LinusU on September 22, 2015 17:7_

@mpdehaan I would to have this implemented, would the behaviour outlined by @candlerb be accepted?

**Comment by candlerb** [comment]
_From @LinusU on September 22, 2015 17:10_

Also, I think that `sudo apt-get install wine=1.2.3` will still work, hold will just prevent it from being upgraded with `dist-upgrade`, `upgrade`, etc...

**Comment by candlerb** [comment]
_From @dmuysson on November 11, 2015 20:51_

+1
We are also using the workaround proposed by @candlerb but it's problematic. You always have to ignore errors as trying to hold an already held package throws errors (as does unholding a package which isn't held). 
Additionally, the other issue noted where --force fails to update a held package has halted more than one playbook in the past, leading to even more steps before an apt-update to try and unhold (potentially) held packages.

In short, supporting the ability to hold/unhold packages via the apt module would lead to cleaner playbooks, and the ability to upgrade held packages via --force would help prevent playbook failure.

**Comment by candlerb** [comment]
_From @weevilgenius on January 5, 2016 2:35_

With versions of apt-mark which support the "showhold" command, you can make holds idempotent with something like this:

```
- command: apt-mark showhold
  register: held_packages
  changed_when: false

- command: apt-mark hold cassandra
  when: '"cassandra" not in held_packages.stdout'
```

**Comment by candlerb** [comment]
_From @mfriedenhagen on January 8, 2016 21:19_

We use a different approach by specifying the version in `/etc/apt/preferences.d/fixed-casandra` like:

```
- copy:
    dest: /etc/apt/preferences.d/fixed-casandra
    content: |
      Package: casandra
      Pin: release v={{casandra_version}}
      Priority: 1002
- apt:
    name: cassandra
    state: latest
```

**Comment by candlerb** [comment]
_From @dgellow on May 13, 2016 17:51_

@mfriedenhagen Thank you for your example. 
Just a correction, instead of 

```
      Pin: release v={{casandra_version}}
```

I had to use

```
      Pin: version {{casandra_version}}
```

**Comment by candlerb** [comment]
_From @mfriedenhagen on May 13, 2016 19:10_

The good thing about the file solution: you may fixate the repository to use as well and may run an occasional apt-get distupgrade without hell breaking loose _and_ no command output has to be parsed.

**Comment by candlerb** [comment]
_From @2rs2ts on July 12, 2016 20:10_

Even if you don't support `apt-mark` in general it would still be nice to give the `apt` module something like the `yum` module's `exclude` option which could be implemented by using `apt-mark` behind the scenes.

**Comment by candlerb** [comment]
_From @ansibot on September 13, 2016 2:33_

@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by candlerb** [comment]
_From @ansibot on October 6, 2016 16:50_

@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by candlerb** [comment]
_From @ansibot on December 7, 2016 16:59_

This repository has been locked. All new issues and pull requests should be filed in https://github.com/ansible/ansible

Please read through the <a href="http://docs.ansible.com/ansible/dev_guide/repomerge.html#move-issues-and-prs-to-new-repo">repomerge</a> page in the dev guide. The guide contains links to tools which automatically move your issue or pull request to the ansible/ansible repo.

**Comment by candlerb** [comment]
Moved from #8060 to https://github.com/ansible/ansible-modules-core/issues/19 and now back

**Comment by TobiMarg** [comment]
I just had this problem and tried the pinning workaround, which worked but only after changing `Priority` to `Pin-Priority`.

**Comment by TimJDFletcher** [comment]
Full working example for reference from Ubuntu 16.04 and docker:
```
- name: Pin Docker-engine release
  copy:
    dest: /etc/apt/preferences.d/docker-engine
    content: |
      Package: docker-engine
      Pin: version {{ docker_engine_version }}
      Pin-Priority: 1002
```

**Comment by geekofalltrades** [comment]
Since Ansible 2.0, you can use dpkg_selections for this: https://docs.ansible.com/ansible/dpkg_selections_module.html

**Comment by mfriedenhagen** [comment]
I do not think this is a proper replacement for fixating the version. You would have to disable the hold, upgrade and hold again. This is much more complicated and error prone  than replacing the version in a preference file.

**Comment by Vladimir-csp** [comment]
Some time ago I suggested a different approach for apt module: https://github.com/ansible/ansible-modules-core/issues/2386

**Comment by ansibot** [comment]
cc @mgwilliams
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by cvernooy23** [comment]
+1

**Comment by samdoran** [comment]
Closing as this capability is available in the `dpkg_selections` module.

**Comment by kpfleming** [comment]
After coming across this issue when searching for a solution to the same problem (I want to block a package from being installed, when insttalling another package which 'recommends' it but does not require it), I tried the `dpkg_selections` method.

Unfortunately `dpkg` intentionally avoids reporting 'hold' state for any packages which are not installed, so if you use the `dpkg_selections` method to set the state of an uninstalled package to 'hold', every time the play is run `dpkg_selections` will report that the status was *changed*, when it was not. If you have other logic in your playbook which cares about this, or even if you're just monitoring the output of ansible-playbook to ensure that no unexpected changes occurred, this is bothersome.

`apt-mark`, on the other hand, properly reports the 'hold' status of uninstalled packages.
